offset.go

```diff
@@ -0,0 +1,414 @@
+}
+
+func offsetParseArgs() {
+	envBrokers := os.Getenv("KT_BROKERS")
+	if config.offset.args.brokers == "" {
+		if envBrokers != "" {
+			config.offset.args.brokers = envBrokers
+		} else {
+			config.offset.args.brokers = "localhost:9092"
+		}
+	}
+	config.offset.brokers = strings.Split(config.offset.args.brokers, ",")
+	for i, b := range config.offset.brokers {
+		if !strings.Contains(b, ":") {
+			config.offset.brokers[i] = b + ":9092"
+		}
+	}
+
+	re, err := regexp.Compile(config.offset.args.topic)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Invalid regex for filter. err=%s\n", err)
+		os.Exit(2)
+	}
+	config.offset.topic = re
+
+	config.offset.partition = int32(config.offset.args.partition)
+
+	config.offset.group = config.offset.args.group
+
+	if config.offset.args.set == "" {
+		config.offset.set = -1
+	} else if config.offset.args.set == "newest" {
+		config.offset.setNewest = true
+	} else if config.offset.args.set == "oldest" {
+		config.offset.setOldest = true
+	} else if newOffset, err := strconv.Atoi(config.offset.args.set); err == nil {
+		config.offset.set = int64(newOffset)
+	} else {
+		fmt.Fprintf(os.Stderr, "Invalid value for setting the offset. possible values are \"oldest\", \"newest\", or any numerical value. err=%s\n", err)
+		os.Exit(2)
+	}
+
+	config.offset.verbose = config.offset.args.verbose
+	config.offset.version = kafkaVersion(config.offset.args.version)
+}
+
+func offsetRun(closer chan struct{}) {
+	var err error
+	if config.offset.verbose {
+		sarama.Logger = log.New(os.Stderr, "", log.LstdFlags)
+	}
+
+	conf := sarama.NewConfig()
+	conf.Version = config.offset.version
+	u, err := user.Current()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to read current user err=%v", err)
+	}
+	conf.ClientID = "kt-offset-" + u.Username
+	if config.offset.verbose {
+		fmt.Fprintf(os.Stderr, "sarama client configuration %#v\n", conf)
+	}
+
+	client, err := sarama.NewClient(config.offset.brokers, conf)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to create client err=%v\n", err)
+		os.Exit(1)
+	}
+	defer client.Close()
+
+	om, err := sarama.NewOffsetManagerFromClient(config.offset.group, client)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to create offset manager for [%s] err=%v\n", config.offset.group, err)
+		os.Exit(1)
+	}
+
+	getBroker := client.Coordinator
+	getPartitions := client.Partitions
+	getOffset := client.GetOffset
+	getPartitionOffsetManager := om.ManagePartition
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	out := make(chan string)
+	go func(out chan string) {
+		topics, err := client.Topics()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to read topics err=%v\n", err)
+			os.Exit(1)
+		}
+
+		for _, t := range topics {
+			if config.offset.args.topic == "" || len(config.offset.topic.FindString(t)) > 0 {
+				offsetsForTopic(getBroker, getPartitions, getOffset, getPartitionOffsetManager, config.offset.group, t, out)
+			}
+		}
+
+		wg.Done()
+	}(out)
+
+	// print to console
+	go func() {
```

Comment:
i think this way you have a chance that messages won't be printed. how about blocking on this for (rather than spawning a goroutine) and quitting it when you close the `out` channel (and close the out channel where you currently call `wg.Done`)?